active/InformationBar/InformationBar.md

```diff
@@ -398,6 +398,62 @@ Please view the guidance for
 for more information about text localization in your notification.
 
 
+# InfoBarPanel class
+
+A layout panel that positions its children horizontally if there’s available space,
+otherwise positions them vertically. This panel is intended to only be used as part
+of the ControlTemplate of the InfoBar control.
+
+## InfoBar Padding properties
+
+The `PaddingInHorizontalOrientation` property gets/sets the distance between the edges
+of the panel and its children, when the panel is laying out items horizontally.
+The `PaddingInVerticalOrientation` property does likewise when the panel is layout out
+items vertically.
+
+[API note: these padding properties are analogous to ex the 
+[StackPanel.Padding](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.StackPanel.Padding)
+property.]
+
+## InfoBar Spacing attached properties
+
+The `SpacingInHorizontalOrientation` attached property gets/sets the distance between i tems
```

Comment:
This is ambiguous; for example if I have two items like this:

```
<Button IB.SIHO="5,0"/>
<Button IB.SIHO="10,0"/>
```
Then is the distance between the two 5, 10, or 15 pixels? I assume it is 15.

Reply:
The distance between the two is 10, because first button has 0 right spacing, and the second button has 10 left spacing.

Reply:
"5,0" means 5 left and right, 0 top and bottom...[?](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Thickness?view=winrt-19041)

Reply:
Looking at the code, this is my understanding of how these are used. 

![image](https://user-images.githubusercontent.com/28935693/100757561-5beee200-33a3-11eb-963c-96e83d201228.png)

Reply:
So the InfoBarPanel contains InfoBar items?

Reply:
Its an implementation detail in the template of InfoBar, it contains the icon, message text etc.. Its just a panel to ensure that all children get the correct layout positions.

Reply:
Couple of things to note
We cannot use SIVO[0].Top instead of PIVO.Top - because there are cases where item0 might be collapsed/hidden.
We cannot replace SIVO[x].Left with PIVO.Left - the values can be different for each item to get the correct visual alignment.

Reply:
What if we rename this to something more appropriate, maybe OrientationWrapPanel if that's what the purpose of this template part? Maybe someone needs something similiar to VariableSizedWrapGrid but instead for the entire collection, not individual items. If it's named InfoBarPanel I am not sure how likely it is that someone will know about the capabilities of the control.

Reply:
> What if we rename this to something more appropriate, maybe OrientationWrapPanel if that's what the purpose of this template part? Maybe someone needs something similiar to VariableSizedWrapGrid but instead for the entire collection, not individual items. If it's named InfoBarPanel I am not sure how likely it is that someone will know about the capabilities of the control.

Could it be a control primitive that would make sense for others to use?

InfoBarPanel does make it seem like it is a Panel that contains InfoBar controls

Reply:
From what I understand it's just a StackPanel that is oriented horizontally if there is enough space, and is oriented vertically if there is not enough horizontal space.

Reply:
I think we could simplify this by having:
* Padding applies to all 4 sides
* Spacing similarly applies to all 4 sides, except it's ignored on the leading and trailing edges

The spacing property was originally named "margin", which I found confusing because the items already have a [Margin](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.FrameworkElement.Margin) property, and because this property is defined by the parent panel, so has parallels to the other panel Spacing properties like [StackPanel.Spacing](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.StackPanel.Spacing). But what this review has brought out is that that's pretty confusing, because it's a Thickness struct like a margin, and it has additive behavior like a margin.

So it looks to me like the padding can just be like every other padding, and the margins like the other Margin property, just two special things to learn:
* This margin property is additive with the existing Margin property
* The leading/trailing margins are ignored. For example, in vertical layout, the top margin of the first visible item and the bottom margin of the last visible item are ignored.

Reply:
Another option -- just throwing it out there -- is to use the `Spacing` concept but allow multiple values. So for example:

```
<InfoBar HorizontalSpacing="5">...
```
In the above snippet with a single attribute value, there is 5 pixels between every item in the control (plus whatever margin the items have on themselves).

```
<InfoBar HorizontalSpacing="5,10,15">...
```
In the above snippet with multiple comma-separated values, the first gap (between item 1 and 2) is 5 pixels, the second gap (betweeen items 2 and 3) is 10 pixels, etc. again, added to whatever margin the individual items have. If there are more items than spacing values, either set to zero or just repeat.

Same for `VerticalSpacing`.

Each item can still use its own `Margin` property for precise alignment, but this makes it trivial to get consistent spacing and easy to get different spacings and not worry about special rules like "left value is ignored on the first item if it's horizontal" and such.

Reply:
I think what I missed originally in this spec is that the "spacing" values are used for more than spacing between the items; in horizontal layout, the spacing.top is also used (intentionally). That's why I'm liking "margin" now, but simplify it by saying that we use all four values of the Thickness, whether horizontal or vertical.

Reply:
Recommend:
* Keep the Padding property, and always use all 4 members of the struct (Top/Bottom/Right/Left)
* Rename the Spacing (attached) properties to ...Margin
* The only special case to the margin properties is to ignore the first/last. 

By "ignore first/last margin", for example in vertical layout, ignore the top margin of the first _visible_ (not collapsed) item and the bottom margin of the last visible item.

Combine that with robloo's suggestion of removing the "In" naming pattern, we have:
* HorizontalOrientationPadding / VerticalOrientationPadding (set on the parent panel)
* HorizontalOrientationMargin / VerticalOrientationMargin (set on the child elements)